server/sonar-web/src/main/js/apps/settings/utils.js

```diff
@@ -0,0 +1,146 @@
+import { TYPE_PROPERTY_SET, TYPE_BOOLEAN, TYPE_SINGLE_SELECT_LIST, TYPE_PASSWORD } from './constants';
+
+export const DEFAULT_CATEGORY = 'general';
+
+export function getPropertyName (definition) {
+  const key = `property.${definition.key}.name`;
+  return hasMessage(key) ? translate(key) : definition.name;
+}
+
+export function getPropertyDescription (definition) {
+  const key = `property.${definition.key}.description`;
+  return hasMessage(key) ? translate(key) : definition.description;
+}
+
+export function getCategoryName (category) {
+  const key = `property.category.${category}`;
+  return hasMessage(key) ? translate(key) : category;
+}
+
+export function getSubCategoryName (category, subCategory) {
+  const key = `property.category.${category}.${subCategory}`;
+  return hasMessage(key) ? translate(key) : getCategoryName(subCategory);
+}
+
+export function getSubCategoryDescription (category, subCategory) {
+  const key = `property.category.${category}.${subCategory}.description`;
+  return hasMessage(key) ? translate(key) : null;
+}
+
+export function getUniqueName (definition, index = null) {
+  const indexSuffix = index != null ? `[${index}]` : '';
+  return `settings[${definition.key}]${indexSuffix}`;
+}
+
+export function getSettingValue (setting) {
+  if (setting.definition.multiValues) {
+    return setting.values;
+  } else if (setting.definition.type === TYPE_PROPERTY_SET) {
+    return setting.fieldValues;
+  } else {
+    return setting.value;
+  }
+}
+
+export function isEmptyValue (definition, value) {
+  if (value == null) {
+    return true;
+  } else if (definition.type === TYPE_BOOLEAN) {
+    return false;
+  } else {
+    return value.length === 0;
+  }
+}
+
+export function getEmptyValue (definition) {
+  if (definition.multiValues) {
+    return [getEmptyValue({ ...definition, multiValues: false })];
+  }
+
+  if (definition.type === TYPE_PROPERTY_SET) {
+    const value = {};
+    definition.fields.forEach(field => value[field.key] = getEmptyValue(field));
+    return [value];
+  }
+
+  if (definition.type === TYPE_BOOLEAN || definition.type === TYPE_SINGLE_SELECT_LIST) {
+    return null;
+  }
+
+  return '';
+}
+
+export function isDefaultOrInherited (setting) {
+  return !!setting.default || !!setting.inherited;
+}
+
+/**
+ * Check if provided definition is a complex one
+ * Complex definitions are property sets or multi values
+ * @param definition
+ * @returns {boolean}
+ */
+export function isComplexDefinition (definition) {
+  return definition.type === TYPE_PROPERTY_SET || definition.multiValues === true;
+}
+
+function getParentValue (setting) {
+  if (setting.definition.multiValues) {
+    return setting.parentValues;
+  } else if (setting.definition.type === TYPE_PROPERTY_SET) {
+    return setting.parentFieldValues;
+  } else {
+    return setting.parentValue;
+  }
+}
+
+/**
+ * Get and format the default value
+ * @param setting
+ * @returns {string}
+ */
+export function getDefaultValue (setting) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Function has a complexity of 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=javascript%3AFunctionComplexity)